src/main/java/hudson/plugins/gradle/injection/DevelocityComputerListener.java

```diff
@@ -21,29 +24,51 @@ public class DevelocityComputerListener extends ComputerListener {
 
     private static final Logger LOGGER = Logger.getLogger(DevelocityComputerListener.class.getName());
 
-    private final DevelocityInjector injector;
+    private final GradleBuildScanInjection gradleBuildScanInjection;
+    private final MavenBuildScanInjection mavenBuildScanInjection;
+    private final MavenExtensionDownloadHandler mavenExtensionDownloadHandler;
     private final Supplier<InjectionConfig> injectionConfigSupplier;
 
     public DevelocityComputerListener() {
-        this(new DevelocityInjector(), new JenkinsInjectionConfig());
+        this(
+                new GradleBuildScanInjection(),
+                new MavenBuildScanInjection(),
+                new MavenExtensionDownloadHandler(),
+                new JenkinsInjectionConfig()
+        );
     }
 
     @VisibleForTesting
-    DevelocityComputerListener(DevelocityInjector injector,
-                               Supplier<InjectionConfig> injectionConfigSupplier) {
-        this.injector = injector;
+    DevelocityComputerListener(
+            GradleBuildScanInjection gradleBuildScanInjection,
+            MavenBuildScanInjection mavenBuildScanInjection,
+            MavenExtensionDownloadHandler mavenExtensionDownloadHandler,
+            Supplier<InjectionConfig> injectionConfigSupplier
+    ) {
+        this.gradleBuildScanInjection = gradleBuildScanInjection;
+        this.mavenBuildScanInjection = mavenBuildScanInjection;
+        this.mavenExtensionDownloadHandler = mavenExtensionDownloadHandler;
         this.injectionConfigSupplier = injectionConfigSupplier;
     }
 
     @Override
     public void onOnline(Computer computer, TaskListener listener) {
         try {
+            InjectionConfig injectionConfig = injectionConfigSupplier.get();
             EnvVars globalEnvVars = computer.buildEnvironment(listener);
-            if (InjectionUtil.globalAutoInjectionCheckEnabled(globalEnvVars) && isFeatureDisabled()) {
+            if (InjectionUtil.globalAutoInjectionCheckEnabled(globalEnvVars) && injectionConfig.isDisabled()) {
                 return;
             }
 
-            injector.inject(computer, globalEnvVars);
+            Map<MavenExtension, String> extensionsDigest = mavenExtensionDownloadHandler.ensureExtensionsDownloaded(
```

Comment:
This can be called concurrently (multiple agents launching at the same time). If somehow the extension was not downloaded after saving the configuration, it could get corrupted when downloaded multiple times. If we think it's a realistic issue we can just have `ensureExtensionDownloaded` making sure the extension is there without downloading it (so no injection in that case). After all now it has to be an explicit action to enable the Maven injection. Wdyt?

Reply:
I've removed downloading on agent's onOnline event as we discussed.

Everything else stays as it's sequential and synchronized by Jenkins